main/src/ca/uwaterloo/flix/language/phase/EarlyTreeShaker.scala

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright 2017 Magnus Madsen, 2022 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.Ast.Annotation
+import ca.uwaterloo.flix.language.ast.TypedAst._
+import ca.uwaterloo.flix.language.ast.{Symbol, TypedAst}
+import ca.uwaterloo.flix.util.Validation._
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps, Validation}
+
+/**
+  * The Tree Shaking phase removes all unused function definitions.
+  *
+  * A function is considered reachable if it:
+  *
+  * (a) The main function is always reachable.
+  *
+  * (b) A function marked with @benchmark or @test is reachable.
+  *
+  * (c) Appears in a function which itself is reachable.
+  *
+  * (d) Is an instance of a class whose signature(s) appear in a reachable function.
+  * Monomorph will erase erase unused instances so this phase must check all instances
+  * for the monomorph to work.
```

Comment:
Maybe @mlutze can help come up with a more concise way of explaining why we should include stuff for the monomorph phase?